x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBytes.java

```diff
@@ -197,11 +208,34 @@ public static void copyToCacheFileAligned(
             final int remainder = buf.position() % PAGE_SIZE;
             final int adjustment = remainder == 0 ? 0 : PAGE_SIZE - remainder;
             buf.position(buf.position() + adjustment);
+            final long writeStartTime = System.nanoTime();
             bytesCopied += positionalWrite(fc, fileChannelPos + bytesCopied, buf);
+            totalWriteTimeNanos += System.nanoTime() - writeStartTime;
```

Comment:
Hm, `System.nanoTime()` is somewhat expensive to call. Ideally we should be using `org.elasticsearch.threadpool.ThreadPool.CachedTimeThread#relativeTimeInNanos` instead. But at the same time it has granularity of about 200ms by default (if I am not mistaken) and it might not be enough to accurately measure this.

Reply:
This makes me think that we should measure the total bytes and convert them to `bytes/aggregation time bucket` when displaying in dashboards.

Reply:
The things we're measuring (copying a buffer) should take nowhere near 200ms, so I don't think `relativeTimeInNanos` is going to be useful here.

We did use `nanoTime()` reliably in my previous role (in low-latency systems, measuring things that ran in the order of single-digit microseconds), but the workloads were typically single threaded so perhaps we never ran into [issues due to contention/monotonicity](https://shipilev.net/blog/2014/nanotrusting-nanotime/).

Does `bytes/aggregation time bucket` give us any meaningful numbers? I assume we could look at the delta in the counter over a time slice, but that would only tell us overall throughput, rather than throughput for individual copies.

If we add a shard ID as an attribute to the metric, (I assume) we could at least get throughput at a bytes-per-shard-per-machine-per-time-slice level, so we might get close to a useful number, but how granular are the time-slices, if the buckets are multiple seconds or more then the throughput we could calculate would potentially get diluted away by the wide bucket size? (i.e. does 1mb/second indicate a slow read or just not much read?)

Reply:
Perhaps it's enough just to calculate the overall throughput (e.g. elapsed time / total bytes written). That would require much fewer calls to `nanoTime`, or perhaps the 200ms granularity of `relativeTimeInNanos` would be sufficient over the time taken to copy a whole chunk? We'd lose the ability to distinguish between the read and write component though.

Reply:
@idegtiarenko I am not 100% sue what you mean by `bytes/aggregation time bucket`. Besides total bytes, what other metrics we need to capture for it? Is there an example for this?

I agree that mutliple calls to `nanoTime` looks a bit concerning. Since the write buffer is 2mb, we could invoke `nanoTime` 32 (8*4) times? I get the intention is to differentiate read and write stats which is something I wanted as well. That said, as a start, it is probably OK to begin with the overall stats, i.e. total bytes (as long counter) and  i.e. total elapsed time in ns (as histogram). These should tell how much data we download and how slow/fast this process is compared to the initial opening stream (we have always assumed opening dominates the latency).

I personally like to record throughput (bytes/elapsed time) as well. With bytes and elapsed time captured separately as suggested in the previous paragraph, we may be able to compute it for selected percentiles at plotting time. But I think it will be likely be rather coarse grained and not very useful. If we assume almost always 16MB data for warming, we can sorta use elapsed time as an approixmation. Another thing to consider is that the current histogram instrument has an odd behaviour of capping at 131,072 which may limit our option for recording througput as histogram. Overall I slightly prefer recording the througput.